### ISSUE TITLE
Headless mode is missing a way to configure rules

### ISSUE THREAD

**Issue by camchenry** [oxc-project/tsgolint#51]
The headless mode is currently working well for enabling rules on a per-file basis. However, there isn't any way to configure rules on a per-file basis (and yes, configuration would likely need to be specified for each file, because it can change depending on glob patterns in the config to apply enable/disable different rules and configurations).

I suggest we allow passing `rules` as an object, similar to the `.oxlintrc.json` or `.eslintrc` format. The keys in `rules` are the rule names, such as `no-floating-promises`. The values in the `rules` object are the configurations. If a rule name is not present as a key, assume it is disabled by default. If an empty object is passed, use the default configuration options.

For example:

```json
{
  "files": [
    {
      "file_path": "/absolute/path/to/file.ts",
      "rules": {
        "no-floating-promises": {
          "allowForKnownSafeCalls": [
            { "from": "file", "name": "safe", "path": "input.ts" }
          ]
        },
      },
    }
  ]
}
```

**Comment by auvred** [comment]
The problem is that Go doesn't have nice enums like Rust does, so the options schema differs slightly from the one used by typescript-eslint. For example, [typescript-eslint's `TypeOrValueSpecifier`](https://typescript-eslint.io/packages/type-utils/type-or-value-specifier/) is a discriminated union, whereas [tsgolint's `TypeOrValueSpecifier`](https://github.com/oxc-project/tsgolint/blob/42e870894061f83bac1358267f7d35c7e670f7c2/internal/utils/type_matches_specifier.go#L22-L29) is just a struct containing all relevant fields.

Ideally, the configuration definition should be written once, and the Go code, Rust code, and documentation should be generated automatically from that single source. Maybe JSON Schema could work?

**Comment by camchenry** [comment]
> The problem is that Go doesn't have nice enums like Rust does, so the options schema differs slightly from the one used by typescript-eslint. For example, [typescript-eslint's `TypeOrValueSpecifier`](https://typescript-eslint.io/packages/type-utils/type-or-value-specifier/) is a discriminated union, whereas [tsgolint's `TypeOrValueSpecifier`](https://github.com/oxc-project/tsgolint/blob/42e870894061f83bac1358267f7d35c7e670f7c2/internal/utils/type_matches_specifier.go#L22-L29) is just a struct containing all relevant fields.

Unfortunately, I have no idea what to suggest here as I'm unfamiliar with Go 😅 It sounds like maybe we need to write a configuration parser that can parse the structs from arbitrary JSON.



> Ideally, the configuration definition should be written once, and the Go code, Rust code, and documentation should be generated automatically from that single source. Maybe JSON Schema could work?

Might be a good idea. Synchronizing across repositories will be difficult, but maybe at least within `tsgolint` we could use codegen to write some of the config parsing code for us? We have a number of scripts in `oxc` that write any tedious implementations over many different shapes of data.

**Comment by Scalahansolo** [comment]
I don't think configuring these rules works at all. 

Defining my rules like so...

```
{
  "rules": {
    "typescript/no-floating-promises": [
      "error",
      { "ignoreVoid": true, "ignoreIIFE": true }
    ],
    "typescript/no-misused-promises": [
      "error",
      {
        "checksVoidReturn": false,
        "checksConditionals": true
      }
    ]
  }
}
```

I am still getting errors about voided returns

**Comment by Boshen** [comment]
We'll need to redefine the schema.

**Comment by camc314** [comment]
To reduce the amount of data we are sending from oxlint -> tsgolint, we are going to switch around how we're storing this info.

Currently the object looks something like:

```ts
type Payload = {
    files: ConfigForFile[]
}
type ConfigForFile = {
    file_path: string
    rules: string
}
```

When linting vscode repo, with all type aware rules enabled, this causes us to pass in `6 712 400` bytes (6.7 MB) via stdin, and this size will only grow when we pass in rule configurations to tsgolint as well.

Since typical projects will be running the same set of rules on all files, we can re-structure the object to look like:

```ts
type Payload = {
    configs: Config[]
}
type Config = {
    file_paths: string[]
    rules: { name: string, config: any }[]
}
```

Which should reduce the duplication of passing in massive JSON objects.

**Comment by camchenry** [comment]
Total agree on changing the schema. A few additional thoughts on ways we could shrink the input even more possibly:

- Use integers for the rules instead of strings. All of the type aware rules could easily fit into a single byte (u8). We would just need to define an enum translation on the oxlint and tsgolint side. This would save quite a few bytes for each rule, especially if many rules are configured.
- Define a base path property of each config. Include the base directory of the tsconfig file for each directory, and then trim each path inside the config to be relative. If there are any paths outside of the base directory, leave it as absolute? Not sure that can happen or not. This would save lots of duplicate bytes in the common case where all paths have the same prefix.

**Comment by Sysix** [comment]
> Not sure that can happen or not. This would save lots of duplicate bytes in the common case where all paths have the same prefix.

Are the file paths absolute?
`oxlint` has a `cwd` where the file searcher will start from. I think this is safe to strip the `cwd` and pass it into `Payload`.

**Comment by theoludwig** [comment]
Related issue #135

**Comment by osdiab** [comment]
another related issue: https://github.com/oxc-project/oxc/issues/13500

**Comment by connorshea** [comment]
Also ran into this problem on the codebase at work, with `typescript/no-misused-promises` not respecting the `checksVoidReturn` option currently. 

https://typescript-eslint.io/rules/no-misused-promises/#checksvoidreturn

**Comment by camchenry** [comment]
## Update

I have a working implementation for configuring rules in headless mode via JSON here: https://github.com/oxc-project/tsgolint/pull/258. It adds a new `options` property to the linting payload for each rule name which can then be parsed by each rule.

It uses `go-jsonschema` to generate the Go structs and unmarshaling code from a JSON Schema file. The JSON schema file is the source of truth for all of the options that a lint rule accepts, and sits alongside the lint rule code in a `schema.json` file. This should ensure that it's fairly easy to modify options and we won't have to maintain as much of the JSON parsing boilerplate code (unless we want to, for performance).

There is probably much room for improvement in the code, but it does work and I think makes it easy to add support for configuring rules. The next step is to implement support on the `oxlint` side, and then complete the configuration support for every rule in `tsgolint`.

**Comment by camchenry** [comment]
oxlint integration is coming along, and I've just completed the first end-to-end test with sending the JSON config to `tsgolint`. From the user perspective, it should be pretty much seamless, it just works(tm).

https://github.com/user-attachments/assets/65a05210-f33a-49fb-9274-67b28a0110b8

From the developer perspective, we'll need to implement some manual JSON serialization/deserialization code for each lint rule. However, it's pretty much normal for other lint rules anyway, plus we need to implement the config support in order to get auto-generated docs so there's not much reason leave everything up to `tsgolint`. Handling config parsing in `oxlint` also makes it easier to handle all errors there rather than needing to return an additional kind of diagnostic from `tsgolint`.

**Comment by Perdolique** [comment]
@camchenry, I see that some changes were merged to `v0.6`. Is it something that already works on the `oxlint` side with this `tsgolont` version and can be tested, or does it require some additional effort somewhere? I can assume the next official release will be alpha. 🤔

**Comment by camchenry** [comment]
@Perdolique No, there isn't anything user-facing that we support which can be used to test this quite yet. The next tsgolint and oxlint versions should contain everything needed to start configuring rules though. We are working on merging a few PRs on the oxlint side to enable rule configuring for tsgolint. There will be an official alpha version available soon.

**Comment by theoludwig** [comment]
I can confirm that using tsgolint v0.7.1 and oxlint v1.29.0, configuring the `no-floating-promises` with the `ignoreVoid` configuration works correctly.

However, it seems like that `allowForKnownSafeCalls` still doesn't work. (#135)
